File: 069/eu069.py
#!/usr/bin/python
# encoding: utf-8
"""
project euler (projecteuler.net) problem 69
solution by Kevin Retzke, April 2012

inefficient solution, max found so far:
300300 57600 5.21354166667
"""
import math

def gcd(n,m):
    """
    Calculate greatest common divisor of n and m using the 
    Euclidian algorithm.
    """
    if n > m:
        large = n
        small = m
    else:
        large = m
        small = n
    if n==0 or m==0:
        return large
    return gcd(large%small,small)

def phi(n):
    result = 0
    for k in xrange(1,n,2):
        if gcd(k,n) == 1:
            result += 1
    return result

if __name__ == '__main__':
    #assert map(phi, range(2,11)) == [1,2,2,4,2,6,4,6,4]
    result = 0
    maxnphin = 0
    for n in xrange(2,10000,2):
        phin = phi(n)
        if phin > 0:
            nphin = float(n)/phin
            if nphin > maxnphin:
                print n, phin, nphin
                result = n
                maxnphin = nphin
    print result

File: 069/eu069.go
// project euler (projecteuler.net) problem 69
// solution by Kevin Retzke (retzkek@gmail.com), May 2012
package main

import (
	"fmt"
	"math"
)

type Primes struct {
	Primes []int
	Last   int
}

// Init initializes a Primes struct with the first two primes.
func (p *Primes) Init() {
	p.Primes = []int{2, 3}
	p.Last = 3
}

// Next computes, returns, and appends the next prime number.
func (p *Primes) Next() int {
	next := 0
	i := p.Last + 2
	for next == 0 {
		sqrti := math.Sqrt(float64(i))
		isPrime := true
		for _, p := range p.Primes {
			if i%p == 0 {
				isPrime = false
				i += 2
				break
			}
			if float64(p) > sqrti {
				break
			}
		}
		if isPrime {
			next = i
		}
	}
	p.Primes = append(p.Primes, next)
	p.Last = next
	return next
}

func main() {
	p := new(Primes)
	p.Init()
	for p.Last < 1e6 {
		p.Next()
	}
    maxn := 0
    maxnphin := float64(0)
    for n := 2; n <= 1000000; n+=2 {
        phi := float64(n)
        for _,m := range p.Primes {
            if n%m == 0 {
                phi *= (1-1/float64(m))
            }
        }
        if nphin := float64(n)/phi; nphin > maxnphin {
            maxn = n
            maxnphin = nphin
            fmt.Println(maxn, maxnphin)
        }
    }
}
